pkg/build/docker/docker.go

```diff
@@ -0,0 +1,367 @@
+package docker
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	docker "github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/jsonmessage"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/kelda/blimp/cli/util"
+	"github.com/kelda/blimp/pkg/auth"
+	"github.com/kelda/blimp/pkg/build"
+	"github.com/kelda/blimp/pkg/errors"
+)
+
+type client struct {
+	client       *docker.Client
+	regCreds     auth.RegistryCredentials
+	dockerConfig *configfile.ConfigFile
+	blimpToken   string
+
+	composeImageCache map[string]types.ImageSummary
+}
+
+type CacheOptions struct {
+	Disable     bool
+	ProjectName string
+}
+
+func New(regCreds auth.RegistryCredentials, dockerConfig *configfile.ConfigFile, blimpToken string, cacheOpts CacheOptions) (build.Interface, error) {
+	dockerClient, err := getDockerClient()
+	if err != nil {
+		return nil, err
+	}
+
+	c := client{
+		client:       dockerClient,
+		regCreds:     regCreds,
+		dockerConfig: dockerConfig,
+		blimpToken:   blimpToken,
+	}
+
+	if !cacheOpts.Disable {
+		composeImageCache, err := getComposeImageCache(dockerClient, cacheOpts.ProjectName)
+		if err == nil {
+			c.composeImageCache = composeImageCache
+		} else {
+			log.WithError(err).Debug("Failed to get compose image cache")
+		}
+	}
+
+	return c, nil
+}
+
+func (c client) BuildAndPush(images map[string]build.BuildPushConfig) (pushedImages map[string]string, err error) {
```

Comment:
I think I had a note that the conn was closing during prepush, did you see that?

Reply:
https://github.com/kklin/blimp-cli/commit/4a67d56c08c0d96e1b83d96731e1f89f3e2a4bda#diff-6e22bcf18906a7e9c0e12b8bdf8146f0R23-R27
I couldn't repro. The connection should have a 30s keepalive ping.